internal/balancer/balancer_test.go

```diff
@@ -0,0 +1,87 @@
+package balancer
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/test/bufconn"
+
+	"github.com/ydb-platform/ydb-go-sdk/v3/config"
+)
+
+func TestBalancer_discoveryConn(t *testing.T) {
+	fakeListener := bufconn.Listen(1024 * 1024)
+	fakeServer := grpc.NewServer()
+	defer fakeServer.Stop()
+
+	go func() {
+		_ = fakeServer.Serve(fakeListener)
+	}()
+
+	var dialAttempt uint32
+
+	balancer := &Balancer{
+		address: "ydbmock:///mock",
+		driverConfig: config.New(
+			config.WithEndpoint("mock"),
+			config.WithGrpcOptions(
+				grpc.WithResolvers(&mockResolverBuilder{}),
+
+				grpc.WithContextDialer(
+					// The first dialing is very long and ends with an error, while the subsequent ones work fine.
+					func(ctx context.Context, s string) (net.Conn, error) {
+						atomic.AddUint32(&dialAttempt, 1)
+
+						if atomic.LoadUint32(&dialAttempt) == 1 {
+							time.Sleep(1 * time.Hour) // extremely slow dialing
+
+							return nil, fmt.Errorf("fake error for endpoint: %s", s)
+						}
+
+						return fakeListener.DialContext(ctx)
+					}),
+
+				// If you want to reproduce the issue, uncomment the line:
+				// grpc.WithDefaultServiceConfig(`{"loadBalancingPolicy": "pick_first"}`),
+			),
+		),
+	}
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 1*time.Second)
```

Comment:
```suggestion
	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
```

Why do you need a second?
Second is a lot of time, then you run a lot of tests. Can you decrease the timeout or use some signals for sync events instead of time (prefer)?

Reply:
I couldn't find any events for synchronization, assuming that there might be retries with different logic and their own timeouts inside `discoveryConn`. If not now, then in the future. One second excludes false test failures in case the test is run on very slow workers. Moreover, one second is only lost if the test fails, and in that case, losing it is not critical. Upon successful completion of the test, the context is canceled via `cancel`.